sympy/polys/orthopolys.py

```diff
@@ -72,10 +72,24 @@ def dup_chebyshevt(n, K):
     """Low-level implementation of Chebyshev polynomials of the first kind."""
     if n < 1:
         return [K.one]
-    m2, m1 = [K.one], [K.one, K.zero]
-    for i in range(2, n+1):
-        m2, m1 = m1, dup_sub(dup_mul_ground(dup_lshift(m1, 1, K), K(2), K), m2, K)
-    return m1
+    # When n is small, it is faster to directly calculate the recurrence relation.
+    if n < 64: # The threshold serves as a heuristic
+        # T_n = 2xT_{n-1} - T_{n-2}
+        m2, m1 = [K.one], [K.one, K.zero]
+        for _ in range(n - 1):
+            m2, m1 = m1, dup_sub(dup_mul_ground(dup_lshift(m1, 1, K), K(2), K), m2, K)
+        return m1
+    # When n is large, the following relation is used for fast computation
+    # T_{2n} = 2T_n^2 - 1
+    # T_{2n+1} = 2T_{n+1}T_n - x
+    m2, m1 = [K.one, K.zero], [K(2), K.zero, -K.one] # T_1, T_2
+    for i in bin(n)[3:]:
+        c = dup_sub_term(dup_mul_ground(dup_mul(m1, m2, K), K(2), K), K.one, 1, K)
+        if  i  == '1':
+            m2, m1 = c, dup_sub_ground(dup_mul_ground(dup_sqr(m1, K), K(2), K), K.one, K)
+        else:
+            m2, m1 = dup_sub_ground(dup_mul_ground(dup_sqr(m2, K), K(2), K), K.one, K), c
+    return m2
```

Comment:
It would be better to make separate functions for each algorithm and then have e.g.:
```
if n < 64:
    return dup_chebyshevt_small(...)
else:
    return dup_chebyshevt_large(...)
```
Perhaps better names can be chosen. Having separate functions is useful so that we can compare their outputs.

Reply:
Created `_dup_chebyshevt_rec` and `_dup_chebyshevt_prod`. I am unsure about the appropriateness of the names.